sys/kern/kern_intr.c

```diff
@@ -531,7 +531,7 @@ intr_event_destroy(struct intr_event *ie)
 {
 
 	if (ie == NULL)
-		return (EINVAL);
+		return (0);
```

Comment:
This change is wrong. We want this to be an error. It's a programming error to call this with NULL.

Reply:
```
% find sys -type f -print0 | xargs -0 grep -eintr_event_destroy -l
sys/netpfil/pf/if_pfsync.c
sys/netpfil/pf/pf_ioctl.c
sys/netpfil/pf/pflow.c
sys/kern/kern_intr.c
sys/kern/subr_intr.c
sys/sys/interrupt.h
% 
```
The ones in `sys/netpfil/pf/` are unlikely to trigger this.  The only present way for them to trigger this is for something to scribble zeros over memory or some other sort of memory corruption.  My concern is they're all calling `intr_event_destroy()` during shutdown, then either doing `MPASS(rc == 0);` or returning the error to the caller.  If this isn't a debug kernel then the error could simply be lost (`sys/netpfil/pf/if_pfsync.c:vnet_pfsync_uninit()`).  If this is a debug kernel, is a `panic()` truly a desirable result?  I suspect _these_ callers would instead prefer:
```
if (ie == NULL) {
        printf("ERROR: %s(): called with NULL event\n", __func__);
        return (0);
}
```
As the rest of their shutdowns can still occur.

The situation for INTRNG (`sys/kern/subr_intr.c`) is very different.  INTRNG does lazy event allocation.  After `intr_isrc_register()` returns, `->isrc_event` is still `NULL`.  `->isrc_event` doesn't become non-`NULL` until `isrc_add_handler()` is called.  As such:
```
struct intr_irqsrc *isrc;
int rc;
intr_isrc_register(isrc);
rc = intr_isrc_deregister(isrc);
printf("intr_isrc_deregister() returned: %d\n", rc);
```
Will reliably output `intr_isrc_deregister() returned: 22`.  So there is a disagreement in the implementation about what `intr_event_destroy(NULL)` should do/return.  Since I'm looking at a driver which needs `intr_isrc_deregister()` to work, this is blocking for me.

I suppose _you_ could modify `isrc_event_destroy()` to do `if (isrc->isrc_event != NULL) return(0);`.  My concern with this is you're testing the same value twice and giving exactly opposite results for a problematic value.  I cannot abide this, it is just too ugly (I will hold my nose and ignore it if someone else implements, but I will not implement).

Another viable approach would be to remove lazy event allocation from INTRNG.  I don't think lazy event allocation significantly helps INTRNG.  I also observe major downsides to lazy event allocation (both here and other places you must first test `->isrc_event`, you cannot assume non-`NULL`).  I think it was a mistake to allow PowerPC and INTRNG to use lazy event allocation.
Issue is this would basically invalidate the premise behind `INTR_SOLO` and someone would be rather annoyed with that (`D35607`).

Reply:
Actually, one other approach _I_ would be willing to implement.  It would be acceptable to me to revert 39888ed7a3dca81c5d5c886e699dbcf58953137f and then have `isrc_event_destroy()` check for `isrc->isrc_event == NULL`.  It is the double checking for `NULL` which I have a problem with.

I though would be quite happy if lazy event allocation was deprecated, and assured removing that would be accepted.

Reply:
In different terms, this situation is highly similar to c1287a3bb1b7b7ccc494d6718db52cb443bd0d38,  Yes, it might be an error, but handing the error off to the caller drastically escalates the damage.  Whereas handling it mitigates things.